Arm/Arm.cs

```diff
@@ -0,0 +1,290 @@
+                var Trans2 = TransformPoint(X, Y);
+                g.DrawLine(Pens.Red, Trans1.Item1, Trans1.Item2, Trans2.Item1, Trans2.Item2);
+                PrevX = X;
+                PrevY = Y;
+            }
+            var Trans = TransformPoint(PrevX, PrevY);
+            g.FillEllipse(Brushes.DarkRed, Trans.Item1 - 3, Trans.Item2 - 3, 6, 6);
+        }
+
+        private (float X, float Y, float Z) ForwardKinematics()
+        {
+            float X = 0.0f;
+            float Y = 0.0f;
+            float Z = 0.0f;
+            float PitchSum = 0.0f;
+            float RollSum = 0.0f;
+            float YawSum = 0.0f;
+            for (int i = 0; i < CurrentPitches.Length; i++)
+            {
+                RollSum += CurrentRolls[i];
+                PitchSum += CurrentPitches[i];
+                YawSum += CurrentYaws[i];
+
+                RollSums[i] = RollSum;
+                PitchSums[i] = PitchSum;
+                YawSums[i] = YawSum;
+
+                X += (float)(Params[i].Length * Math.Cos(PitchSum) * Math.Cos(YawSum));
+                Y += (float)(Params[i].Length * Math.Cos(RollSum) * Math.Sin(PitchSum));
+                Z += (float)(Params[i].Length * Math.Sin(RollSum) * Math.Sin(YawSum));
+            }
+            return (X, Y, Z);
+        }
+
+        private (float, float, float) CalcGrad(float TX, float TY, float TZ, float X, float Y, float Z, int i)
+        {
+            float RollY = 0.0f;
+            float RollZ = 0.0f;
+            float PitchX = 0.0f;
+            float PitchY = 0.0f;
+            float YawX = 0.0f;
+            float YawZ = 0.0f;
+
+            for (int a = i; a < Params.Length; a++)
+            {
+                RollZ += (float)(Params[a].Length * Math.Cos(RollSums[a]) * Math.Sin(YawSums[a]));
+                RollY += (float)(Params[a].Length * Math.Sin(RollSums[a]) * Math.Sin(PitchSums[a]));
+
+                PitchX += (float)(Params[a].Length * Math.Sin(PitchSums[a]) * Math.Cos(YawSums[a]));
+                PitchY += (float)(Params[a].Length * Math.Cos(PitchSums[a]) * Math.Cos(RollSums[a]));
+
+                YawX += (float)(Params[a].Length * Math.Sin(YawSums[a]) * Math.Cos(PitchSums[a]));
+                YawZ += (float)(Params[a].Length * Math.Cos(YawSums[a]) * Math.Sin(RollSums[a]));
+            }
+            return (2 * RollZ * (TZ - Z) - 2 * RollY * (TY - Y),
+                    2 * PitchX * (TX - X) - 2 * PitchY * (TY - Y),
+                    2 * YawX * (TX - X) - 2 * YawZ * (TZ - Z));
+        }
+
+        private static float EnsureStandardForm(float Angle)
+        {
+            while (Angle < 0)
+                Angle += (float)(2 * Math.PI);
+            return Angle % (float)(2 * Math.PI);
+        }
+
+        public void MoveTo(float X, float Y, float Z)
+        {
+            const float LearningRate = 0.0001f;
+            const int Iterations = 40;
+            for (int j = 0; j < Params.Length; j++)
```

Comment:
I think this loop and the one that follows are swapped. We should be adjusting each parameter on every iteration.

Reply:
I’ve noticed that this way performs much better because it updates the most impactful angles first. We can test it further, but from my preliminary testing it looked like this is better.

Reply:
I remembered the main rationale for me doing this while writing the code: it lets us clamp the angle after the gradient descent, meaning that the angle can move through illegal territory during gradient descent and only get clamped afterwards.